lib/internal/modules/package_json_reader.js

```diff
@@ -154,10 +175,50 @@ function getPackageType(url) {
   return getPackageScopeConfig(url).type;
 }
 
+const pjsonImportAttributes = { __proto__: null, type: 'json' };
+let cascadedLoader;
+/**
+ * @param {URL['href'] | URL['pathname']} specifier The location for which to get the "root" package.json
+ * @param {URL['href'] | URL['pathname']} parentLocation The location of the current module (ex file://tmp/foo.js).
+ */
+function findPackageJSON(specifier, parentLocation) {
+  validateString(specifier, 'specifier');
+
+  let parentURL;
+  if (path.isAbsolute(parentLocation)) {
+    parentURL = `${pathToFileURL(parentLocation)}`;
+  } else {
+    validateURLString(parentLocation, 'parentLocation');
+    parentURL = parentLocation;
+  }
+
+  let resolvedTarget;
+  if (specifier[0] === '.') {
+    resolvedTarget = `${new URL(specifier, `${path.dirname(parentURL)}/`)}`;
+  } else {
+    cascadedLoader ??= require('internal/modules/esm/loader').getOrInitializeCascadedLoader();
+
+    try {
+      resolvedTarget = cascadedLoader.resolve(specifier, parentURL, pjsonImportAttributes).url;
```

Comment:
Let's try to not convert to a string too soon, as it ends up complexifying the code (and making it slower). Also, let's avoid `path.` methods as they can be mutated in userland and are OS-dependent (i.e. the Windows version won't work on URL strings)
```suggestion
  const parentURL = URLParse(parentLocation) || pathToFileURL(parentLocation);

  let resolvedTarget;
  if (specifier[0] === '.') {
    resolvedTarget = new URL(specifier, parentURL);
  } else {
    cascadedLoader ??= require('internal/modules/esm/loader').getOrInitializeCascadedLoader();

    try {
      resolvedTarget = cascadedLoader.resolve(specifier, `${parentURL}`, pjsonImportAttributes).url;
```

Reply:
Is `URL.canParse` the same as `URLParse`?